### PR TITLE
SONARJAVA-2292 Add NOT_PRESENT constraint to value returned by empty()

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java [hunk 1]
```diff
@@ -62,6 +66,21 @@ public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
...
+  public ProgramState checkPostStatement(CheckerContext context, Tree syntaxNode) {
+    List<ProgramState> programStates = setNotPresentConstraint(context, syntaxNode);
+    Preconditions.checkState(programStates.size() == 1);
+    return programStates.get(0);
+  }
+
+  private static List<ProgramState> setNotPresentConstraint(CheckerContext context, Tree syntaxNode) {
+    SymbolicValue val = context.getState().peekValue();
+    if(syntaxNode.is(Tree.Kind.METHOD_INVOCATION) && OPTIONAL_EMPTY.matches(((MethodInvocationTree) syntaxNode))) {
+      return val.setConstraint(context.getState(), OptionalConstraint.NOT_PRESENT);
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java]: MAJOR A "NullPointerException" could be thrown; "val" is nullable here. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java [hunk 2]
```diff
@@ -62,6 +66,21 @@ public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
...
 
+  @Override
+  public ProgramState checkPostStatement(CheckerContext context, Tree syntaxNode) {
+    List<ProgramState> programStates = setNotPresentConstraint(context, syntaxNode);
+    Preconditions.checkState(programStates.size() == 1);
+    return programStates.get(0);
+  }
+
+  private static List<ProgramState> setNotPresentConstraint(CheckerContext context, Tree syntaxNode) {
+    SymbolicValue val = context.getState().peekValue();
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java]: `val` can be moved within the `then` clause of the `if` statement.

File: java-frontend/src/test/files/se/ConditionAlwaysTrueOrFalseCheck.java [hunk 3]
```diff
@@ -2123,3 +2123,14 @@ void bar(A a) {
     }
   }
 }
+
+class OptionalEmptyNotPresent {
+  private Optional fun() {
+    java.util.Optional<String> op = java.util.Optional.empty();
+    if (op.isPresent()) { // Noncompliant
```
Comment (m-g-sonar) [inline java-frontend/src/test/files/se/ConditionAlwaysTrueOrFalseCheck.java]: I would add the message of the issue, to confirm it's an "always false", and not an "always true"

File: java-frontend/src/test/files/se/ConditionAlwaysTrueOrFalseCheck.java [hunk 4]
```diff
@@ -2123,3 +2123,14 @@ void bar(A a) {
     }
   }
 }
+
+class OptionalEmptyNotPresent {
```
Comment (m-g-sonar) [inline java-frontend/src/test/files/se/ConditionAlwaysTrueOrFalseCheck.java]: I would also add a test on the `OptionalGetBeforeIsPresentCheck` test file to show that we are raising issue explicitly on them as well.

File: java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java [hunk 5]
```diff
@@ -62,6 +66,20 @@ public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
...
+  @Override
+  public ProgramState checkPostStatement(CheckerContext context, Tree syntaxNode) {
+    List<ProgramState> programStates = setNotPresentConstraint(context, syntaxNode);
+    Preconditions.checkState(programStates.size() == 1);
+    return programStates.get(0);
+  }
+
+  private static List<ProgramState> setNotPresentConstraint(CheckerContext context, Tree syntaxNode) {
+    if(syntaxNode.is(Tree.Kind.METHOD_INVOCATION) && OPTIONAL_EMPTY.matches(((MethodInvocationTree) syntaxNode))) {
+      return context.getState().peekValue().setConstraint(context.getState(), OptionalConstraint.NOT_PRESENT);
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java]: MAJOR A "NullPointerException" could be thrown; "peekValue" is nullable here. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

**Pull request merged**